x-pack/plugins/task_manager/task_poller.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/*
+ * This module contains the logic for polling the task manager index for new work.
+ */
+
+import { Logger } from './lib/logger';
+
+type WorkFn = () => Promise<void>;
+
+interface Opts {
+  pollInterval: number;
+  logger: Logger;
+  work: WorkFn;
+}
+
+/**
+ * Performs work on a scheduled interval, logging any errors. This waits for work to complete
+ * (or error) prior to attempting another run.
+ */
+export class TaskPoller {
+  private isStarted = false;
+  private isWorking = false;
+  private timeout: any;
+  private pollInterval: number;
+  private logger: Logger;
+  private work: WorkFn;
+
+  /**
+   * Constructs a new TaskPoller.
+   *
+   * @param opts
+   * @prop {number} pollInterval - How often, in milliseconds, we will run the work function
+   * @prop {Logger} logger - The task manager logger
+   * @prop {WorkFn} work - An empty, asynchronous function that performs the desired work
+   */
+  constructor(opts: Opts) {
+    this.pollInterval = opts.pollInterval;
+    this.logger = opts.logger;
+    this.work = opts.work;
+  }
+
+  /**
+   * Starts the poller. If the poller is already running, this has no effect.
+   */
+  public async start() {
+    if (this.isStarted) {
+      return;
+    }
+    this.isStarted = true;
+
+    const poll = async () => {
+      await this.attemptWork();
+
+      if (this.isStarted) {
+        this.timeout = setTimeout(poll, this.pollInterval);
+      }
+    };
+
+    poll();
+  }
+
+  /**
+   * Stops the poller.
+   */
+  public stop() {
+    this.isStarted = false;
+    clearTimeout(this.timeout);
+    this.timeout = undefined;
+  }
+
+  /**
+   * Runs the work function. If the work function is currently running,
+   * this has no effect.
+   */
+  public async attemptWork() {
+    if (this.isWorking) {
+      return;
+    }
+
+    this.isWorking = true;
+
+    try {
+      await this.work();
+    } catch (error) {
+      const message = error.stack || error.toString();
```

Comment:
Can you elaborate? Because it seems ok to me:
![image](https://user-images.githubusercontent.com/908371/49758816-23691180-fc7d-11e8-826c-de12fdbe8d66.png)

Reply:
Change `return Promise.reject('i like turtles');` to

- `return Promise.reject();`
- `return Promise.reject(undefined);`
- `return Promise.reject(null);`

Reply:
Interpolating an Error object in a string will automatically get the message string from the error. I'll just do that. I don't think we truly need the stacktrace, because we have the `Failed to cancel task` prefix message - we'll find what's going wrong from that.

Reply:
589da32b67